reinstate no-spawn option for remove

dwsutherland opened the pull request:
At present there's no way to remove parentless tasks from the scheduler and not have them spawn their next instance.

This is not ideal, and at ESNZ I've run into this problem many times while adding new tasks to a workflow and them being spawned in at the distant past (i.e. the ICP), and then the only way to fix the workflow is to spawn them forward one cycle at a time to the desired cycle point.

Also, in general, perhaps you wish not to run an isolate/branch for a gap then introduce again at some future cycle point.

This PR fixes the problem by reinstating `--no-spawn` option to `cylc remove`..

Example:
```
[scheduling]
    initial cycle point = 20260101T0000Z
    [[graph]]
        P1D = """
            @wall_clock => a => b => c
            d
        """

[runtime]
    [[root]]
        script = sleep 5
    [[a,b,c,d]]
```
running shows:
<img width="310" height="211" alt="image" src="https://github.com/user-attachments/assets/dd2033bf-f2b8-4689-8b89-3be36af5f1c3" />
Removing `d` normally (`$ cylc remove couple/run1//20260106T0000Z/d`) spawns it into the next cycle:
<img width="306" height="206" alt="image" src="https://github.com/user-attachments/assets/e756b1a6-5272-43f4-ba10-f9b63ee88239" />
With `--no-spawn`:
<img width="305" height="162" alt="image" src="https://github.com/user-attachments/assets/256838d9-903f-47bc-a66a-4a7df669682f" />

And of course removing the sequentially spawned `a` (`cylc remove --no-spawn couple/run1//20260101T0000Z/a`), with this option, gives:
<img width="317" height="138" alt="image" src="https://github.com/user-attachments/assets/7334f3af-f765-45e6-822d-e89acde3f890" />
(as expected)

Code changes were made in such a way that they should be conflict-free with #7132 

Seems our forms are smart enough at the UI end:
<img width="490" height="447" alt="image" src="https://github.com/user-attachments/assets/2a77b548-d491-4a27-8d6a-2abdcc85aa7b" />


**Check List**

- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [x] Applied any dependency changes to both `setup.cfg` (and `conda-environment.yml` if present).
- [x] Tests are included (or explain why tests are not needed).
- [x] Changelog entry included if this is a change that can affect users
- [x] [Cylc-Doc](https://github.com/cylc/cylc-doc) pull request opened if required at cylc/cylc-doc/pull/XXXX. (not needed I think, [docs](https://cylc.github.io/cylc-doc/stable/html/user-guide/interventions/index.html#remove-tasks) aren't that detailed about removal and options are explained in UI/CLI info)
- [x] If this is a bug fix, PR should be raised against the relevant `?.?.x` branch.

dwsutherland committed:
reinstate no-spawn option for remove

dwsutherland committed:
tests added

dwsutherland committed:
change log entry

hjoliver commented:
>  I've run into this problem many times while adding new tasks to a workflow and them being spawned in at the distant past (i.e. the ICP),

How did they get spawned back at the ICP - hopefully that does not happen automatically?

(Even if that was human error, it's a still a problem that we need to be able to fix easily, but just wondering...)

oliver-sanders commented:
As it stands, the `cylc remove` command is a bit odd, it can do one of three things:

1. Remove a single instance of a task.
2. Temporally remove a single instance of a task (if it gets re-spawned as the result of another output).
3. Remove all instances of all tasks (i.e, stop the workflow) or lop off an entire branch of the workflow.

All depending on the exact details of the graph structure and the precise state of the task pool at the time the command was issued. The user can't really tell what will happen without a lot of graph understanding and SoD knowledge which is a shame.

The new option might be a little bit confusing (i.e, it only applies to parentless / sequential-xtriggererd tasks), but the status quo is a bit confusing, so whatever.

@hjoliver, thoughts?

dwsutherland commented:
> Yes, however, the flip side of the flip side is that the remove solution requires you to know enough about the graph structure

This:
> we may still need some lower-level commands or options for expert use in some situations

Also, it's not an option slapping users in the face.. It's there if needed.

> Another option might be to set the tasks

Problem here is you run into the same skip-broadcast issue; you may inadvertently trigger a cascade of downstream tasks to run/spawn..
This is one main reason why the `--no-spawn` option is useful..

oliver-sanders commented:
To be clear, not arguing against the use case here, and we do need some kind of a solution to enable the removal of orphan tasks (#7209).

This argument is absolutely an option, but:

* It might not work reliably:
   * E.g, It's not preserved across restarts / reloads, we can handle this of course.
  * But it's going to introduce a bunch more interactions for us to test, e.g, how does this interact with set, trigger, plain-remove, plain-restart, auto-restart, graph-changes, etc.
* It will most likely only ever be used by one person!
  * It's [really hard to explain what `cylc remove` does](https://github.com/cylc/cylc-flow/pull/7134#issuecomment-3950432581) in the first place
    * It's completely contextual, so its behaviours can't be easily predicted ahead of time.
    * In the extreme, removing just one task could cause an entire infinitely-cycling workflow to shut down!
  * It's really really hard to explain the situations in which this new option would be needed in the help text.
    * This involves explaining parentless and sequentially-xtriggered tasks.
    * As well as explaining that sequential xtriggers may be implicit, e.g, as in the case of `@wall_clock`.
  * It's really really really hard for an operator to determine whether an incoming workflow configuration change is such a situation:
    * They need to diff the before/after graphs, then identify all tasks which are parentless OR triggered by a sequential xtrigger.
  * It's really really really really hard for an operator to compose an intervention to mitigate against this.
    * They need to identify the link tasks which join the old tasks onto the new ones, note these dependencies may differ from cycle to cycle.
  * Odds are that there is only one operator in the world with the insight required to use this feature.
    * Impressive skills, but I feel there should be a less involved way to achieve this!

Hence trying to explore whether there are other options here. If we can't come up with a better idea...

In the SoS days, the remove option had spawn/no-spawn modes which worked consistently in all scenarios, whereas as it stands here, the absence of `--no-spawn` does not infer that future instances of the task will be spawned (status quo remove behaviour).

<br />

>> Another option might be to set the tasks
>
> Problem here is you run into the same skip-broadcast issue; you may inadvertently trigger a cascade of downstream tasks to run/spawn..

This plays both ways, with the remove approach, you may inadvertently trigger a cascade of downstream tasks if:
* You fail to identify [all of] the correct task(s) to remove.
* You fail to use the `--no-spawn` option.

With the set approach, you may inadvertently trigger a cascade of downstream tasks if:
* You fail to identify [all of] the correct task(s) to set.

The set approach would be a bit harder to get wrong because we can automatically generate the list of newly added tasks (and already log this for reloads), whereas we cannot automatically determine the correct task(s) to remove (which may be different from one cycle to another).

Moreover, with the remove approach, you may create the conditions for a workflow stall due to missing task outputs, whereas the set and skip approaches ensure that this cannot happen.

dwsutherland commented:
> Would it help to construct a dummy workflow...

A basic example would be to create a gap, i.e.  remove without spawning at one cycle and reintroduce at some future cycle..
The tests here just about do that, would just need a reintroduction (`cylc set --pre=all  . . .`)..

hjoliver commented:
Thanks @oliver-sanders - I'll attempt to find the time to digest your write-up in detail later today.

As a first comment, I was preferring the remove option as much simpler in terms of both implementation and command syntax (if perhaps not conceptually simpler, being a low level intervention) - you only need to target a single task.

> cylc set //<2027/{a,b,c} - (theoretical cycle selection syntax)

Your other suggestions rely on selecting a potentially large number of tasks over many cycles and modifying all of those in the DB. Maybe that's OK though, with the right syntax (I like your suggested syntax there).

dwsutherland commented:
Bear with me, I do come to an agreement in the end. . .

> **Caveats to the remove approach**
> 
> To conclude this lengthy debate, several caveats to the remove approach have been identified.

If you don't like it from an idealistic point of view, that's fine, however these points are off target.. 
I'm not looking for something to replace `set/skip`, or some general solution or conceptual reframing (although maybe you are), I'm looking for a solution to particular scenarios where `set/skip` are less practical (scenarios common at ESNZ).. 


> The following major points have been conceded (by acknowledgement or lack of counter):
> 
> 1. The remove operation needs to be performed on the whole subgraph **in the general case** in order to reliably remove these tasks
>    
>    * But not in your special case where all added tasks in a single chain.

This point is void.. I've explained already:
- I would **not** use it on the "whole subgraph" just the parentless beginning of a sub-graph.
- The special case is the usual case.

> 2. The remove operation may need to be performed multiple times
>    
>    * In the likely event that not all downstream tasks spawn simultaneously.

I object here too.. To create/avoid a gap I would usually just target the parentless task(s) of a cycle/glob-of-cycle point(s) if necessary. 


> 3. Remove can only mop up tasks after they have spawned (and potentially after they have submitted / started).
>    
>    * Note that your approach of pausing the workflow to avoid this will only work if there is no inter-dependence between the R1 tasks, otherwise the workflow will have to be resumed to allow them to run.

Yes, this is a post spawn intervention.. Yes, it would be nice to avoid the situation in the first place, but people make mistakes.. And it's nice that they can visualize the thing they need to act on (as opposed to the "whole subgraph" which they would have to know about.

> 4. The same tasks will have to be re-removed with any subsequent re-running of the R1 tasks.
>    
>    * Really surprised you're not concerned about this! From your description, this applies to your use case!
> 
> Additionally, this point was not conceded but was demonstrated in a worked example:

I already addressed this:

> I would `set` the first downstream R1 task(s), in this case, and `remove` the any downstream and next instance spawned...
This would have to out weigh the trade offs of using `set` here and future cycles...

That's why I'm not concerned..
 
> 5. Unintended active tasks may occur.
>    
>    * The logical result of (3) above, proven in example.

And again, I've already addressed this.. There are "unintended active tasks" with `skip/set` approaches, and I would say it's more pernicious.. I would rather things don't kick off because of fake succeeded states in the workflow..


---



> **General case solution**
> 
> > If you want to convince me to not have remove --no-spawn as an option,
> 
> I'm not necessarily trying to convince you not to have `remove --no-spawn` as an option (read back, I noted that this actually necessary, at least internally, for #7198), I'm pointing out that it's not a great solution and doesn't work for all cases, there is more likely than not a better resolution to this scenario.
> 
> 
> > you'll need to show me that there is a more practical way to handle these examples.
> 
> Great, let's talk alternatives!
> 
> Outline of the scenario:
> 
> * We have run a bunch of cycles.
> * We have added some tasks, and reloaded/restarted the workflow.
> * We have triggered an explicit selection of tasks somewhere in the workflow's history.
> * We expected only the explicitly triggered tasks to spawn/run, but some other tasks got spawned too.
> * (note that this problem is not coupled to R1 tasks, though this is a pressing use case at present)
> 
> Outline of the problem:
> 
> * When we trigger a historical task, it _may_ cause newly added tasks to spawn and either run (if dependencies are satisfied) or force a workflow-stall depending on the specifics of the graph change.
> 
> 
> While the status-quo behaviour is "logical" by SoD, it's rather illogical from a user perspective. It can certainly be argued that this is a bug (I lean towards the position that it is, more on that below).
> 
> Irrespective of whether we accept this as a bug or not, there are two possible ways for Cylc to handle this situation:
> 
> 1. Spawn these tasks.
> 2. Don't spawn these tasks.
> 
> We have to choose one of these as the default behaviour, currently, we choose (1), however, I would argue that:
> 
> * (2) is a more logical default.
> * Not spawning these tasks is the most likely user intention.
> * The inverse intervention (i.e, to spawn these tasks) is easier and less caveat-prone this way around.
> 

I'm with you..

> 
> But it would be much easier to prevent this from happening in the first place. We can always provide a mechanism for running these new-old tasks if anyone were ever to want this.

Agree, again..


> This problem is actually very similar to the issue of pre-start-cycle-point tasks in the warm start scenario (#7178), an issue which is impacting our operational workflows at the moment:
> 
> * With warm starts, all tasks are newly added, we presume that all historical task instances (tasks before the warm-start point) have completed. This is the so-called "pre-initial" logic.
> * With this scenario, it is arguable that we should consider all historical task instances (tasks before the reload-point) as completed.

There's another issue I've run into, which this will solve, and that's off ICP start tasks (when you spin up a workflow on specified tasks, cycles ahead of the ICP)..
You can accidentally end up spawning in these tasks earlier when running R1  (as this gap is meaningless after a reload/restart)..  Setting earlier instances as complete would help with this. 


> So there's two simple alternatives:
> 
> * Consider historical instances of newly-added tasks to be completed.
> * Set the start cycle point of the recurrences of newly-added tasks to the oldest active cycle point.
> 
> Both of these approaches would completely resolve this situation, removing the need for the operator to run the `cylc remove --no-spawn` command.

**Agreed!**
Also, to remove gaps and alleviate the extent to which these alternatives need to play a part:
I want isolated-graphs/inter-graph-dependencies to be a thing:
https://github.com/cylc/cylc-flow/issues/7020#issuecomment-3644811990

Along with ideas around parentless tasks https://github.com/cylc/cylc-flow/issues/7228   (which I agree that xtriggers should be like automatic python tasks separate from what they trigger)


**For now:**

This PR is a small change that can be undone once we have some of these solutions Ironed out and ready to go (which I assume would be something like `8.8` release).

dwsutherland commented:
I might give this a crack/go.

> We would probably need to find an alternative way to implement group-trigger to make this approach handle the use case here.

WRT this -- I feel that, post reload, if triggering a group (say `FAM1`) contains new tasks, then they should be targeted (as the group encompasses the new tasks)...  

However, you're correct, the way it's implemented now will mean off-group new tasks could also spawn in... So perhaps we could either:
- Remember the previous record before wiping it (in a temporary dict or something), and check it before spawning off shoots of this group..
- Or internally create a temporary flow or pass/ticket for the group, that gives them an exemption to run (as the previous record).

The 2nd option seems cleaner; the temporary pass/one-off-ticket implementation could by-pass the record-checking system (that determines whether a task should run or not), and the group ticket could be handed down to dependents when the corresponding outputs are produced (if they're in the group specified on the ticket).
For reload/restart this ticket would have to be recorded in the DB against the current ticket holders.

hjoliver commented:
I just ran into a nasty problem myself for which `remove --no-spawn` is the best solution: I added (via a colleague who vouched for it) a large number (several hundred!) of mostly independent tasks to our operational "downstream sync" workflow, to sync operational ensemble output filesets to our other site, then manually kicked off the new tasks with bulk trigger commands. A lot failed due to bad file matching globs, others succeeded even though they should have failed (because evidently rsync failure due to ssh connection limiting (for DDOS protection!) doesn't return error status), and another lot were in the wrong offset recurrence (our ensemble has different tasks in 03, 15 vs 09, 21).

The easiest way out of this total nightmare would have been to `remove` (via a small number of globbed commands) all the new instances that ran (and succeeded or failed) and all those that had spawned ahead and were waiting in future cycles - i.e. to back out of the runtime change to make it look as if it had never happened (then fix the config and trigger the corrected stuff). But I could not do that because removing the waiting instances spawns the next ones, ad infinitum.

Stopping the workflow to warm start it wasn't practical because of the spread over cycles and the sheer number of reruns that would cause. I actually restarted it with my own hacked scheduler (with david's change) to allow remove to do what I wanted - which worked like a charm 💯

dwsutherland commented:
Well.. I think this one is still ready to go, Oliver may have mentioned something about restart/reload  (from memory), but maybe that was for something else because this would work for your case and they wouldn't spawn back on reload/restart (only when triggered manually or automatically via outputs)

hjoliver commented:
Yes, `remove` makes it look as if the target task had never spawned. There's no reload/restart issue, at least for this sort of use case. After removing a task, it *should* spawn again if upstream circumstances demand that - that's the whole point of remove: (a) get rid of  the current (or past) instance; and (b) allow it run again (if needed) as if it had never spawned in the first place.

(I've posted a comment on Element)

hjoliver commented:
Note after doing the above, I realized I can set runahead limited waiting instances to `failed` and then remove them, and it doesn't spawn the next instance - but that's actually a bug, it *should* spawn the next instance under normal circumstances!  Is it a known bug...?  (And in any case, it is more difficult and less intuitive than `remove --no-spawn`).

[Update: #7413]

oliver-sanders commented:
> I just ran into a nasty problem myself for which remove --no-spawn is the best solution
>
> The easiest way out of this total nightmare would have been to remove (via a small number of globbed commands) all the new instances that ran (and succeeded or failed) 

Yes, `cylc remove` seems like the natural solution to this problem, ideally we wouldn't *need* a `--no-spawn` option for this operation to work correctly.

For the ICP/trigger scenario (which started this off), we have agreed that it would be *better* if these tasks were never spawned in the first place (avoiding the need for any subsequent remove operation) - this is an option.

<br />

> Stopping the workflow to warm start it wasn't practical

IMO: Warm start is an emergency operation, a great capability for Cylc to have as an ultimate backstop, but not something which should be considered for regular interventions.

dwsutherland commented:
Partially addressing PR up: https://github.com/cylc/cylc-flow/pull/7419

oliver-sanders commented:
Can we close this PR as superseded by https://github.com/cylc/cylc-flow/pull/7419?

hjoliver commented:
Ping @dwsutherland ☝️

dwsutherland commented:
> Can we close this PR as superseded by #7419?

Yes, will do a separate thing for the no spawn on re-outputs thing..
